controller/api/public/top_routes.go

```diff
@@ -2,40 +2,207 @@ package public
 
 import (
 	"context"
+	"errors"
 	"fmt"
 	"sort"
 	"strings"
 
+	sp "github.com/linkerd/linkerd2/controller/gen/apis/serviceprofile/v1alpha1"
 	pb "github.com/linkerd/linkerd2/controller/gen/public"
+	api "github.com/linkerd/linkerd2/controller/k8s"
 	"github.com/linkerd/linkerd2/pkg/k8s"
 	"github.com/prometheus/common/model"
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
 )
 
 const (
 	routeReqQuery             = "sum(increase(route_response_total%s[%s])) by (%s, dst, classification)"
 	routeLatencyQuantileQuery = "histogram_quantile(%s, sum(irate(route_response_latency_ms_bucket%s[%s])) by (le, dst, %s))"
-	dstLabel                  = `dst=~"%s(:\\d+)?"`
+	dstLabel                  = `dst=~"(%s)(:\\d+)?"`
+	// DefaultRouteName is the name to display for requests that don't match any routes.
+	DefaultRouteName = "[DEFAULT]"
 )
 
+type dstAndRoute struct {
+	dst   string
+	route string
+}
+
+type indexedTable = map[dstAndRoute]*pb.RouteTable_Row
+
+type resourceTable struct {
+	resource string
+	table    indexedTable
+}
+
 func (s *grpcServer) TopRoutes(ctx context.Context, req *pb.TopRoutesRequest) (*pb.TopRoutesResponse, error) {
+	errRsp := validateRequest(req)
```

Comment:
consider a `log.Debugf("TopRoutes request: %+v", req)` at the top of this method.